src/back/space/src/repo/table_space.cpp

```diff
@@ -0,0 +1,278 @@
+#include "table_space.hpp"
+#include "../../../shared/errors.hpp"
+#include <chrono>
+
+#include <userver/components/component_config.hpp>
+#include <userver/components/component_context.hpp>
+#include <userver/utils/boost_uuid4.hpp>
+#include <userver/yaml_config/merge_schemas.hpp>
+#include <userver/storages/postgres/component.hpp>
+
+#include <boost/uuid/uuid_io.hpp>
+
+namespace svetit::space::table {
+
+namespace pg = storages::postgres;
+
+Space::Space(storages::postgres::ClusterPtr pg)
+	: _pg{std::move(pg)}
+{
+	constexpr auto kCreateTable = R"~(
+CREATE TABLE IF NOT EXISTS space (
+	id UUID PRIMARY KEY,
+	name TEXT NOT NULL,
+	key TEXT NOT NULL,
+	requestsAllowed BOOLEAN NOT NULL,
+	createdAt BIGINT NOT NULL
+);
+)~";
+
+	using storages::postgres::ClusterHostType;
+	_pg->Execute(ClusterHostType::kMaster, kCreateTable);
+
+	//InsertDataForMocks();
+}
+
+const storages::postgres::Query kInsertSpace{
+	"INSERT INTO space (id, name, key, requestsAllowed, createdAt) "
+	"VALUES ($1, $2, $3, $4, $5) ",
+	storages::postgres::Query::Name{"insert_space"},
+};
+
+void Space::Insert(
+	const boost::uuids::uuid& uuid,
+	const std::string& name,
+	const std::string& key,
+	const bool requestsAllowed,
+	const int64_t createdAt)
+{
+	_pg->Execute(storages::postgres::ClusterHostType::kMaster, kInsertSpace, uuid, name, key, requestsAllowed, createdAt);
+}
+
+const storages::postgres::Query kSelectSpace{
+	"SELECT id, name, key, requestsAllowed, createdAt "
+	"FROM space OFFSET $1 LIMIT $2",
+	storages::postgres::Query::Name{"select_space"},
+};
+
+std::vector<model::Space> Space::Select(const int offset, const int limit)
+{
+	auto res = _pg->Execute(storages::postgres::ClusterHostType::kMaster, kSelectSpace, offset, limit);
+
+	if (res.IsEmpty())
+	{
+		return {};
+	}
+
+	return res.AsContainer<std::vector<model::Space>>(pg::kRowTag);
+}
+
+const storages::postgres::Query kSelectSpaceAvailable{
+	R"~(
+		SELECT id, name, key, requestsAllowed, createdAt
+		FROM space
+		WHERE requestsAllowed = true
+		AND id NOT IN (
+			SELECT spaceId FROM space_user WHERE userId = $1
+		) OFFSET $2 LIMIT $3
+	)~",
+	storages::postgres::Query::Name{"select_space_available"},
+};
```

Comment:
Давай вместо вложенных select использовать join везде
```suggestion
const storages::postgres::Query kSelectSpaceAvailable{R"~(
	SELECT s.id, s.name, s.key, s.requestsAllowed, s.createdAt
	FROM space s
	LEFT JOIN space_user su ON s.id = su.spaceId
	WHERE s.requestsAllowed = true AND su.userId != $1
	OFFSET $2 LIMIT $3)~",
	storages::postgres::Query::Name{"select_space_available"},
};
```

Reply:
Готово (везде).

Reply:
Кроме https://github.com/svetit-io/svetit/blob/feature/36-ms-spaces-back/src/back/space/src/repo/table_space.cpp#L53 и https://github.com/svetit-io/svetit/blob/feature/36-ms-spaces-back/src/back/space/src/repo/table_space.cpp#L65 - здесь вернул вложенный select, так как не понял, как реализовать аналог NOT IN через JOIN.

Reply:
А предложенный мною вариант в этом комментарии не заработал?

Reply:
Нет, твой вариант именно тут не заработал, так как, если не ошибаюсь вообще, LEFT JOIN срабатывает, как аналог WHERE id IN (SELECT spaceId ...). А здесь обратное условие WHERE id NOT IN (SELECT spaceId ...). Сейчас попрошу chatgpt переписать данный запрос)

Reply:
ChatGPT вот так переписала (вроде, работает)):

`SELECT s.id, s.name, s.key, s.requestsAllowed, s.createdAt
FROM space s
LEFT JOIN space_user su ON s.id = su.spaceId AND su.userId = $1
WHERE s.requestsAllowed = true AND su.spaceId IS NULL
OFFSET $2 LIMIT $3;`

Reply:
Готово.

Reply:
Мда, чот я тут напутал от незнания. :( Твой первый вариант был самый лучший. Извини. Можем оставить вариант ChatGPT или вернуть твой первый вариант, как хочешь.